File: code_samples/haskell/design_patterns_factory.hs
-- Design Patterns: Factory
-- AI/ML Training Sample

module Factory where

data Factory = Factory {
    getData :: String
} deriving (Show, Eq)

process :: Factory -> String -> Factory
process obj input = obj { getData = input }

validate :: Factory -> Bool
validate obj = not (null (getData obj))

main :: IO ()
main = do
    let instance = Factory { getData = "" }
    let updated = process instance "example"
    putStrLn $ "Data: " ++ getData updated
    putStrLn $ "Valid: " ++ show (validate updated)

File: code_samples/go/design_patterns_factory.go
package main

import (
    "fmt"
)

// Design Patterns: Factory
// AI/ML Training Sample

type Factory struct {
    Data string
}

func NewFactory() *Factory {
    return &Factory{
        Data: "",
    }
}

func (s *Factory) Process(input string) {
    s.Data = input
}

func (s *Factory) Validate() bool {
    return len(s.Data) > 0
}

func (s *Factory) GetData() string {
    return s.Data
}

func main() {
    instance := NewFactory()
    instance.Process("example")
    fmt.Printf("Data: %s\n", instance.GetData())
    fmt.Printf("Valid: %v\n", instance.Validate())
}
